internal/services/containers/kubernetes_cluster_resource.go

```diff
@@ -675,6 +676,37 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
 				},
 			},
 
+			"key_vault_kms": {
+				Type:     pluginsdk.TypeList,
+				Optional: true,
+				ForceNew: false,
+				MaxItems: 1,
+				Elem: &pluginsdk.Resource{
+					Schema: map[string]*pluginsdk.Schema{
+						"enabled": {
+							Type:     pluginsdk.TypeBool,
+							Required: true,
+						},
+						"key_id": {
+							Type:         pluginsdk.TypeString,
+							Required:     true,
+							ValidateFunc: keyVaultValidate.NestedItemId,
+						},
+						"key_vault_network_access": {
+							Type:         pluginsdk.TypeString,
+							Default:      string(managedclusters.KeyVaultNetworkAccessTypesPublic),
+							Optional:     true,
+							ValidateFunc: validation.StringInSlice(managedclusters.PossibleValuesForKeyVaultNetworkAccessTypes(), false),
+						},
+						"key_vault_resource_id": {
```

Comment:
This field can be removed since we can extract the `key_vault_resource_id` from the `key_vault_key_id`

Reply:
Sure?
Example Key ID:
`"https://kv-aks-ci-m98ln0012-97lm.vault.azure.net/keys/aks-ci-m98ln0012-97lm5x-etcd-encryption/3b2c3c949bc14771afcc580470861398"`

Example Vault Resource ID:
`"/subscriptions/fb49fd82-eaec-4cb6-a69d-580c5c530263/resourceGroups/rg-aks-ci-m98ln0012-97lm5x/providers/Microsoft.KeyVault/vaults/kv-aks-ci-m98ln0012-97lm"`

Reply:
Sure, we can use the parsing functions available in the [keyvault service package](https://github.com/hashicorp/terraform-provider-azurerm/tree/main/internal/services/keyvault/parse) to parse the key id, from that we should have the necessary information to generate the associated key vault id, so something like this in the expand function:
```
	keyId, err := keyVaultParse.KeyID(raw["key_vault_key_id"].(string))
	if err != nil {
		return err
	}

	keyVaultId := keyVaultParse.NewVaultID(keyId.SubscriptionId, keyId.ResourceGroup, keyId.VaultName)
```

Reply:
Will try this :-). But I am a bit sceptical since the subscription and the resource group is missing in the keyId.

Reply:
Tried it. Exactly as I expected:

```go
func TestAccKubernetesCluster_tryoutKeyVaultParse(t *testing.T) {
	key, err := keyVaultParse.KeyID("https://kv-aks-ci-m98ln0012-97lm.vault.azure.net/keys/aks-ci-m98ln0012-97lm5x-etcd-encryption")
	log.Printf("Key: %v", key)
	log.Printf("Err: %v", err)
}
```

Output:
```log
go test -timeout 30s -run ^TestAccKubernetesCluster_tryoutKeyVaultParse$ github.com/hashicorp/terraform-provider-azurerm/internal/services/containers -v
=== RUN   TestAccKubernetesCluster_tryoutKeyVaultParse
2023/01/11 12:29:56 Key: <nil>
2023/01/11 12:29:56 Err: No subscription ID found in: "keys/aks-ci-m98ln0012-97lm5x-etcd-encryption"
--- PASS: TestAccKubernetesCluster_tryoutKeyVaultParse (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/containers	1.853s
```

Reply:
Taking a step back, does this property even need to be exposed? It's optional and following the tutorial in the MSFT docs it doesn't need to be supplied in the payload in order to configure it. So to me it looks like this can be removed entirely?

Reply:
Ah I see, I think it will improve the user experience and is worth the effort to extract this behind the scenes instead of forcing the user to conditionally supply it depending on the network access mode. Here is an example on how to extract the key vault id from the key id using the vault client:
https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/compute/disk_encryption_set_resource.go#L347-L363

Reply:
I can take another look into this but: Do you really think TF should be different thant the Azure REST API and the Azure CLI? Both of them need that information

Reply:
As you pointed out, the API only requires the key vault id if the network access mode is `Private`, so the API would expect the user to know when and in what scenarios to provide which information. 

Since the key vault id is information that is easily attainable from the key id, I think abstracting that complexity away, (which in turn also minimises the configuration that needs to be supplied as well as less properties to document for a resource that already has a very bloated documentation) will improve the user experience and the provider does do that in a lot of places.

Reply:
I am having a look into :-)

But I am definitively more biased to have a dedicated field for this for simplicity. There are always edge cases if we want to autodetect that. Eg. KeyVault, Cluster identity and Key is managed from another team (different TF state).
But as sait, will have a look into 👍

Reply:
@stephybun Had success, implemented it in [ffca49e](https://github.com/hashicorp/terraform-provider-azurerm/pull/19893/commits/ffca49ea78a7598c077df3b7cde77f33129c4564). Let me now if I can do anything else.
Shall squash the commits into one?